src/WRings.sol

```diff
@@ -0,0 +1,124 @@
+        iRings = iRings_;
+    }
+
+    /// @return assetTokenAddress Contract address of token utilized by the vault
+    function asset() public view returns(address){
+        address assetTokenAddress = address(iRings);
+        return assetTokenAddress;
+    }
+
+    /// @return totalManagedAssets Balance of the asset token within the vault
+    function totalAssets() public view returns(uint256) {
+        uint256 totalManagedAssets = iRings.balanceOf(address(this));
+        return totalManagedAssets;
+    }
+
+    /// @notice Allows the owner to set the exchange rate
+    /// @dev uses 27 decimals of precision, and can be no greater than 1000000000000000000
+    /// @param exchangeRate_ New exchange rate
+    function setExchangeRate(uint256 exchangeRate_) external onlyOwner {
+        require(exchangeRate_ <= 100000000000000000 * 10 ** 27, "New rate is outside bounds");
+        exchangeRate = exchangeRate_;
+    }
+
+    /// @notice The amount of shares to mint to a user for the asset token deposited under ideal conditions
+    /// @param assets Number of Rings tokens to deposit to the contract 
+    /// @return shares Number of vault shares minted to the user based on the exchange rate
+    function convertToShares(uint256 assets) external view returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        
+        return shares;
+    }
+
+    /// @notice Used to calculate the amount of an asset token to transfer to user for vault tokens burned under ideal conditions
+    /// @param shares Number of vault tokens to burn
+    /// @return assets Number of asset tokens to transfer to the user
+    function convertToAssets(uint256 shares) external view returns(uint256) {
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
+        return assets;
+    }
+
+    /// @notice User specifies amount of asset token to deposit in return for vault shares
+    /// @param assets Number of asset tokens to deposit
+    /// @param receiver Address that vault tokens will be minted to
+    /// @return shares Number of vault shares to return to the user
+    function deposit(uint256 assets, address receiver) external  returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        mint(receiver, shares);
+        iRings.transferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return shares;
+    }
+
+    /// @notice User specifies the desired amount of vault shares in exchange for asset tokens
+    /// @param shares Number of vault tokens desired by the user
+    /// @param receiver Address that vault Tokens will be minted to
+    /// @return assets Number of asset tokens required to mint desired amount of shares
+    function mint(uint256 shares, address receiver) external  returns(uint256) {
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
```

Comment:
Instead of repeating math logic, you should call the `convertToAssets` function.

Even better, for gas cost. You should put the logic on an internal `_convertToAssets` function, and call `_convertToAssets` from the external `convertToAssets` (which becomes a user-facing function), `mint` and `redeem`.

The same pattern should be followed for `convertToShares`, `deposit` and `withdraw`.